File: packages/silo-client-java/src/main/java/in/org/quicko/silo/client/errors/ValidationFailedException.java
package in.org.quicko.silo.client.errors;

import java.util.List;

/** A {@code 400}: the request was malformed, with one entry per field the validator rejected. */
public class ValidationFailedException extends SiloException {
  private final List<ValidationDetail> details;

  public ValidationFailedException(String message, String method, String path, List<ValidationDetail> details) {
    super(400, "validation_failed", message, method, path);
    this.details = List.copyOf(details);
  }

  /** One entry per rejected field, or empty when the server named none. */
  public List<ValidationDetail> details() {
    return details;
  }
}

File: packages/silo-client-dart/lib/src/errors/validation_failed_exception.dart
import 'silo_exception.dart';
import 'validation_detail.dart';

/// A `400`, with one detail per rejected field.
final class ValidationFailedException extends SiloException {
  const ValidationFailedException(String message, String method, String path, this.details)
    : super(400, 'validation_failed', message, method, path);

  factory ValidationFailedException.fromWire(String message, String method, String path, Object? details) =>
      ValidationFailedException(message, method, path, [
        if (details is List)
          for (final detail in details)
            if (detail case {'path': final String pointer, 'message': final String text})
              ValidationDetail(pointer, text),
      ]);

  final List<ValidationDetail> details;
}
